lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php

```diff
@@ -4,378 +4,95 @@
-                $this->connections['slave'] = $this->_conn;
-            }
+        if (isset($params['slaves'])) {
+            $this->deprecated("Params key 'slaves'", "'replica'");
 
-            return false;
+            $params['replica'] = $params['slave'];
+            unset($params['slave']);
         }
 
-        if ($connectionName === 'master') {
-            $this->connections['master'] = $this->_conn = $this->connectTo($connectionName);
+        if (isset($params['keepSlave'])) {
+            $this->deprecated("Params key 'keepSlave'", "'keepReplica'");
 
-            // Set slave connection to master to avoid invalid reads
-            if (! $this->keepSlave) {
-                $this->connections['slave'] = $this->connections['master'];
-            }
-        } else {
-            $this->connections['slave'] = $this->_conn = $this->connectTo($connectionName);
+            $params['keepReplica'] = $params['keepSlave'];
+            unset($params['keepSlave']);
         }
 
-        if ($this->_eventManager->hasListeners(Events::postConnect)) {
-            $eventArgs = new ConnectionEventArgs($this);
-            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
-        }
-
-        return true;
+        parent::__construct($params, $driver, $config, $eventManager);
     }
 
     /**
-     * Connects to a specific connection.
-     *
-     * @param string $connectionName
-     *
-     * @return DriverConnection
+     * Checks if the connection is currently towards the primary or not.
      */
-    protected function connectTo($connectionName)
+    public function isConnectedToMaster(): bool
     {
-        $params = $this->getParams();
+        $this->deprecated('isConnectedtoMaster()', 'isConnectedToPrimary()');
 
-        $driverOptions = $params['driverOptions'] ?? [];
-
-        $connectionParams = $this->chooseConnectionConfiguration($connectionName, $params);
-
-        $user     = $connectionParams['user'] ?? null;
-        $password = $connectionParams['password'] ?? null;
-
-        return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
+        return $this->isConnectedToPrimary();
     }
 
     /**
-     * @param string  $connectionName
-     * @param mixed[] $params
+     * @param string|null $connectionName
      *
-     * @return mixed
+     * @return bool
      */
-    protected function chooseConnectionConfiguration($connectionName, $params)
+    public function connect($connectionName = null)
     {
         if ($connectionName === 'master') {
-            return $params['master'];
-        }
-
-        $config = $params['slaves'][array_rand($params['slaves'])];
+            $connectionName = 'primary';
 
-        if (! isset($config['charset']) && isset($params['master']['charset'])) {
-            $config['charset'] = $params['master']['charset'];
+            $this->deprecated('connect("master")', 'connect("primary")');
```

Comment:
this should hint at `ensureConnectedToPrimary()` instead. and I even suggest not supporting `connect('primary')` at all (so that the new class does not need to add a parameter in `connect()` compared to the parent Connection class)